Build common interface for waypoint geometry constraints, migrate IP placement.

DanAlbert opened the pull request:
This is a replacement for the existing "zone geometry" classes that are currently used for choosing locations for IP, hold, and join points. The older approach required the author to define the methods for choosing locations at a rather low level using shapely APIs to merge or mask geometries. Debug UIs had to be defined manually which was a great deal of work. Worse, those debug UIs were only useable for *successful* waypoint placement. If there was a bug in the solver (which was pretty much unavoidable during development or tuning), it wasn't possible to use the debug UI.

This new system adds a (very simple) geometric constraint solver to allow the author to describe the requirements for a waypoint at a high level. Each waypoint type will define a waypoint solver that defines one or more waypoint strategies which will be tried in order. For example, the IP solver might have the following strategies:

1. Safe IP
2. Threat tolerant IP
3. Unsafe IP
4. Safe backtracking IP
5. Unsafe backtracking IP

We prefer those in the order defined, but the preferred strategies won't always have a valid solution. When that happens, the next one is tried.

The strategies define the constraints for the waypoint location. For example, the safe IP strategy could be defined as (in pseudo code):

* At least 5 NM away from the departure airfield
* Not farther from the departure airfield than the target is
* Within 10 NM and 45 NM of the target (doctrine dependent)
* Safe
* Within the permissible region, select the point nearest the departure
  airfield

When a solver fails to find a solution using any strategy, debug information is automatically written in a GeoJSON format which can be viewed on geojson.io.

This also adds some fuzz testing which generates random inputs for waypoint solvers to check if they can find a solution. If they can't, the debug info for the solver is dumped to the testcases directory. Another test loads those test cases, creates a solver from them, and checks that a solution is found. Obviously it won't be immediately, but it's a starting point for fixing the bug and serves as a regression test afterward.

Fixes https://github.com/dcs-liberation/dcs_liberation/issues/3085.

DanAlbert committed:
Build common interface for waypoint geometry constraints.

This is a replacement for the existing "zone geometry" classes that are
currently used for choosing locations for IP, hold, and join points.
The older approach required the author to define the methods for
choosing locations at a rather low level using shapely APIs to merge or
mask geometries. Debug UIs had to be defined manually which was a great
deal of work. Worse, those debug UIs were only useable for *successful*
waypoint placement. If there was a bug in the solver (which was pretty
much unavoidable during development or tuning), it wasn't possible to
use the debug UI.

This new system adds a (very simple) geometric constraint solver to
allow the author to describe the requirements for a waypoint at a high
level. Each waypoint type will define a waypoint solver that defines one
or more waypoint strategies which will be tried in order. For example,
the IP solver might have the following strategies:

1. Safe IP
2. Threat tolerant IP
3. Unsafe IP
4. Safe backtracking IP
5. Unsafe backtracking IP

We prefer those in the order defined, but the preferred strategies won't
always have a valid solution. When that happens, the next one is tried.

The strategies define the constraints for the waypoint location. For
example, the safe IP strategy could be defined as (in pseudo code):

* At least 5 NM away from the departure airfield
* Not farther from the departure airfield than the target is
* Within 10 NM and 45 NM of the target (doctrine dependent)
* Safe
* Within the permissible region, select the point nearest the departure
  airfield

When a solver fails to find a solution using any strategy, debug
information is automatically written in a GeoJSON format which can be
viewed on geojson.io.

Fixes https://github.com/dcs-liberation/dcs_liberation/issues/3085.

DanAlbert committed:
Migrate IP placement to WaypointSolver.

SnappyComebacks commented on tests/flightplan/test_waypointstrategy.py:
```diff
@@ -0,0 +1,167 @@
+
+def test_safe_prerequisite_safe_point() -> None:
+    strategy = WaypointStrategy(MultiPolygon([]))
+    strategy.prerequisite(Point(0, 0)).safe()
```
Perhaps doing something like `to_be_safe()`, `is_safe()`, or`should_be_safe()` or similar instead of `safe()` would be a little more readable?  I wasn't completely certain what this was doing until I read the test `test_safe_prerequisite_unsafe_point()`. (Also thank you for writing tests!)

  I suppose we could also define multiple functions of the same functionality if you feel like there are different contexts that are suited for different ones.  Though that might be a detriment unless they could all live within the same file/area.

SnappyComebacks replied:
Reading the other tests it may be worth changing other functions such `nearest` as well.  I read `test_no_solution_if_prerequisites_failed` as "Find a safe point nearest to {nearest Point} if {prerequisite Point} is safe."

If I seem to be misunderstanding the goal or premise here please let me know.  I think having these functions be close to natural language would be quite helpful.

DanAlbert replied:
+1 to renaming `safe()` to `is_safe()`, done.

DanAlbert replied:
For the second point, I added some comments:

```python
def test_no_solution_if_prerequisites_failed() -> None:
    """Verify that no solution is found if prerequisites are not satisfied.

    This test has a 1-meter radius threat zone about the center of the plane. It has a
    prerequisite for a safe center, which will fail. The test verifies that even if
    there are no .require() constraints that would prevent finding a solution, failed
    prerequisites still prevent it (prerequisites differ from constraints in that they
    will prevent any of the other operations from happening without needing to location
    constraints, which is important because it allows strategies to avoid defending
    against invalid cases).
    """
    strategy = WaypointStrategy(MultiPolygon([Point(0, 0).buffer(1)]))
    strategy.prerequisite(Point(0, 0)).is_safe()
    # This constraint won't actually apply, but it's required before calling find() so
    # we need to set it even though it's not actually relevant to the test.
    strategy.nearest(Point(0, 0))
    assert strategy.find() is None
```

SnappyComebacks commented:
Do you foresee things like waypoint altitude and/or speed being chosen by `Strategy` types as well, or is that a separate concern?

DanAlbert commented:
> Do you foresee things like waypoint altitude and/or speed being chosen by `Strategy` types as well, or is that a separate concern?

I addressed this a bit in discord, but the record here is more useful so I'll repeat myself:

Assuming you mean something like "should `require().safe()` consider 25k ft above a strela as safe?", yes, that'd be cool, but it's not something we can do without knowing where the ground is. The Caucasus map includes both coastal terrain at ~0 ft MSL and Mount Elbrus at 18,510 ft MSL, so without reasonably high fidelity DEMs we can't do that.

Speed, no, I haven't thought of any ways that'd affect waypoint placement. It's more likely we'll fit the speed to the location than the location to the speed anyhow.

SnappyComebacks commented:
Keeping in mind that these were all under the umbrella of "Should strategies manage these things," and not strictly thinking about `is_safe()` - these were the things I was thinking about for altitude and speed:

**Altitude:**
- Weather - under the clouds or over them affects target visibility and AI attack behavior.  That doesn't matter for Strike and JDAMs but it does matter for CAS, BAI and targets of opportunity.  Not only would you prefer to have visibility at target but you would probably want to be able to see while cruising to/from target as well.  Same considerations for Barcap/Tarcap.
- Some planes have much lower effective cruise altitude than others.  (Effective = with a bomb load)
- I suppose safety from ground threats is a concern as well like you mentioned.
- Some weapons do better starting from high altitude than low and close.

**Speed:**
- Should wind be considered?  At worst case, high winds mean flying at a high IAS one way and a low IAS the other way.  (Before the speeds were tuned lower I saw some AI cruising in afterburner to try to meet the required ground speed)

SnappyComebacks commented:
With regards to not having a DEM available, I think we're stuck at the moment with "Are you hugging the ground or at a predefined altitude?"  I suppose we could think about it in the sense of "Here the ground is generally flat/smooth" and "Here the ground is bumpy", and then rely on the AI to do some level of terrain avoidance through flat terrain.

DanAlbert commented:
Those are all important things but I'm not sure what we'd gain by handling them as part of this?

SnappyComebacks commented:
> Those are all important things but I'm not sure what we'd gain by handling them as part of this?

That's ok too.  I was mostly asking to gauge the scope of what you want this branch (and the new features it includes) to accomplish.

DanAlbert commented:
Gotcha. Yeah, my only goal here is the stuff described by the bug: make the existing thing more maintainable, debuggable, and easier for others to write (so far I'm the only one that's contributed any in the existing form, and they take a _lot_ of head scratching).

Altitude, speed, and all kinds of other crap falls under the umbrella of #270 (I'll post my plans once I have them sorted out, but this is a prerequisite for that so I got sidetracked).

DanAlbert committed:
Add fuzz testing for waypoint solvers.

This fuzz test generates random inputs for waypoint solvers to check if
they can find a solution. If they can't, the debug info for the solver
is dumped to the testcases directory. Another test loads those test
cases, creates a solver from them, and checks that a solution is found.
Obviously it won't be immediately, but it's a starting point for fixing
the bug and serves as a regression test afterward.

DanAlbert committed:
Add test cases found by fuzzer.

SnappyComebacks commented:
Something I just thought about.  If I understand correctly we are evaluating the waypoints directly here.  Do we need to evaluate the path between the waypoints as well?  The scenario I just thought of is having a threat zone in the middle of two distant waypoints.

DanAlbert committed:
Force dumping debug info on recreate.

We need a way to debug successful solvers that's still targeting to a
specific flight. This will do for now.

DanAlbert commented:
> Something I just thought about. If I understand correctly we are evaluating the waypoints directly here. Do we need to evaluate the path between the waypoints as well? The scenario I just thought of is having a threat zone in the middle of two distant waypoints.

Yes, but that's not a regression.
